### DISCUSSION TITLE
Airflow integration  with Pager duty

### DISCUSSION THREAD

**Discussion by Praful19** [apache/airflow#21086] [Q&A]
Hi Team,

I have an automation requirement where we need to create incidents in PagerDuty using scripts. whenever AIrflow Dags/Task got failed. Is there any documentation regarding the same?

Which scripting language is recommended to do so?
I was planning for Python or Terraform. Any suggestions on this.

**Comment by josh-fell** [comment]
You could probably use the [PagerdutyEventsHook](https://github.com/apache/airflow/blob/main/airflow/providers/pagerduty/hooks/pagerduty_events.py) to create an incident tickets in a function and the function is set as the `on_failure_callback` for your DAGs.  The `on_failure_callback` can be set at the DAG level so that any task in the DAG that fails triggers the function.  From the [Airflow Python API documentation](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/models/dag/index.html):
> on_failure_callback (callable) – A function to be called when a DagRun of this dag fails. A context dictionary is passed as a single parameter to this function.

You can find more information on callbacks [in this documentation](https://airflow.apache.org/docs/apache-airflow/stable/logging-monitoring/callbacks.html?highlight=callback).

**Reply by Praful19** [reply]
Hi @josh-fell  Thanks for your response it was really helpful however I have few more queries before calling function if you can help me with that would be really great.

1- Where we need to call the function is it on Dag level or Task level ? Or is there any way we can pull the data from metadata so that can be passed to pager duty ? 
2-  if we are calling it on Dag Level do we have to update same function inorder to trigger pagerduty alert if the Dags fail ? 
3- is there any way where we can pull data from metadata or something and pass it pagerduty using event API ? 

Regards
Praful

**Reply by josh-fell** [reply]
> 1- Where we need to call the function is it on Dag level or Task level ? Or is there any way we can pull the data from metadata so that can be passed to pager duty ?

The callback can be set at either the DAG level (which acts as a default for all tasks) or for an individual task. What do you mean by "metadata"? The execution `context` object is passed to the callback function if that's what you mean. You can also do whatever you want in the callback function.


> 2- if we are calling it on Dag Level do we have to update same function inorder to trigger pagerduty alert if the Dags fail ?

Yes the callback is a custom function you write with whatever logic you want to trigger for a given task status. Airflow handles calling the callback for you.


> 3- is there any way where we can pull data from metadata or something and pass it pagerduty using event API ?

Again, what you you mean by "pull data from metadata"?
